add logcat logging and timeout

aerickson opened the pull request:
Debugging branch to test fixes for https://bugzilla.mozilla.org/show_bug.cgi?id=1611936.

#35 didn't reveal if script.py is the source of the hang (https://treeherder.mozilla.org/#/jobs?repo=try&revision=72e4a6def23410d0e371b9fb86a0c62df4e563cc).

Based on #35. Does the following:
- Add logcat logging to get insight into what the command is doing.
- Adds a timeout at 44 minutes. `pstree` will be shown.

```
02-08 06:36:02.361  6420  6420 I log     : script.py: 1581143762.9057271:+0.0016665458679199219: command started
02-08 06:36:12.402  6425  6425 I log     : script.py: 1581143762.9057271:+10.042171716690063: ll:4 bw:44 br:44 rc:None
02-08 06:36:22.445  6429  6429 I log     : script.py: 1581143762.9057271:+20.079461812973022: ll:4 bw:84 br:84 rc:None
02-08 06:36:32.480  6433  6433 I log     : script.py: 1581143762.9057271:+30.119335651397705: ll:4 bw:124 br:124 rc:None
02-08 06:36:42.514  6436  6436 I log     : script.py: 1581143762.9057271:+40.15351867675781: ll:4 bw:164 br:164 rc:None
02-08 06:36:52.546  6439  6439 I log     : script.py: 1581143762.9057271:+50.18122053146362: ll:4 bw:204 br:204 rc:None
02-08 06:37:02.366  6442  6442 I log     : script.py: 1581143762.9057271:+60.006245374679565: timeout at 1 minute(s)
02-08 06:37:02.475  6445  6445 I log     : script.py: 1581143762.9057271:+60.11710524559021: pstree: 
02-08 06:37:02.475  6445  6445 I log     : 'bash,1
02-08 06:37:02.475  6445  6445 I log     :   |-adb,15 -L tcp:5037 fork-server server --reply-fd 7
02-08 06:37:02.475  6445  6445 I log     :   |   |-{adb},16
02-08 06:37:02.475  6445  6445 I log     :   |   |-{adb},22
02-08 06:37:02.475  6445  6445 I log     :   |   |-{adb},23
02-08 06:37:02.475  6445  6445 I log     :   |   |-{client_socket_t},19
02-08 06:37:02.475  6445  6445 I log     :   |   `-{device poll},18
02-08 06:37:02.475  6445  6445 I log     :   `-python3,89 /builds/taskcluster/script.py /bin/bash -c for i in {1..90}; do echo "444"; sleep 1; done ## run script 10 times
02-08 06:37:02.475  6445  6445 I log     :       |-bash,190 -c for i in {1..90}; do echo "444"; sleep 1; done ## run script 10 times
02-08 06:37:02.475  6445  6445 I log     :       |   `-sleep,268 1
02-08 06:37:02.475  6445  6445 I log     :       `-sh,272 -c /usr/bin/pstree -aApct
02-08 06:37:02.475  6445  6445 I log     :           `-pstree,273 -aApct'
02-08 06:37:02.580  6448  6448 I log     : script.py: 1581143762.9057271:+60.22212743759155: command finished: ll:4 bw:240 br:240 rc:None
```

aerickson committed:
ensure sys.stdout.write writes everything

aerickson committed:
add debug to detect underwrites

aerickson committed:
wip

aerickson committed:
wip2

aerickson committed:
add more debug output

aerickson committed:
don't need dupe line

aerickson committed:
indicate command finished

aerickson committed:
add psmisc to docker file (for pstree)

aerickson committed:
add timeout

aerickson committed:
fix pstree output, adjust timeout

aerickson committed:
tweak newlines

aerickson committed:
cleanup

aerickson committed:
tweaks

aerickson committed:
import cleanup

aerickson committed:
get rid of extra newlines

aerickson committed:
ignore .vscode and .idea dirs

aerickson committed:
refactoring/cleanup

aerickson committed:
add vim

aerickson committed:
wip: log timeout to logcat also

aerickson committed:
bug fix

aerickson committed:
remove duplicate stdout prints

aerickson committed:
simplify

aerickson committed:
tweak pstree command

aerickson committed:
grammar

aerickson committed:
cleanup

aerickson committed:
set default rc code to 1

aerickson committed:
tweak pstree outupt

aerickson committed:
pad pstree output

aerickson committed:
try to fix logcat mangling pstree output

aerickson committed:
fix up pstree output

aerickson committed:
set timeout to 44 m

aerickson committed:
no point in changing this default

aerickson committed:
remove newlines

bclary commented on Dockerfile:
```diff
@@ -29,13 +29,15 @@ RUN apt-get update && \
     sudo \
     tzdata \
     unzip \
+    vim \
```
Why do we need vim? Is this for your local development and debugging?

aerickson replied:
Yeah, just for testing locally. Removed.

bclary commented on Dockerfile:
```diff
@@ -29,13 +29,15 @@ RUN apt-get update && \
     net-tools \
     netcat \
     openjdk-8-jdk-headless \
+    psmisc \ 
```
trailing space.

aerickson replied:
Fixed.

bclary commented on taskcluster/script.py:
```diff
@@ -203,22 +261,32 @@ def main():
-    print("script.py: command finished (bytes read: %s, bytes written: %s)" % (bytes_read, bytes_written))
+    dpi = DebugPrinter(device)
+    # timeout in x minutes
+    with timeout(TIMEOUT_MINUTES * 60, dpi):
```
interesting use...

aerickson replied:
> Let's make the <= changes. I definitely don't want this merged to master though. Let's just deploy this branch to a testing image to get the debugging information and we can move on from there. I'd like to schedule a meeting this coming week to talk about this script.py. I think we should consider making a number of changes but I'd like to work more closely with you on them.

Yeah, the previous change wasn't landed to master either.

Meeting sounds good.

bclary commented on taskcluster/script.py:
```diff
@@ -203,22 +261,32 @@ def main():
+            rc = proc.poll()
+            if line:
+                temp_bytes_written = 0
+                while temp_bytes_written != line_len:
```
let's change this to <=

aerickson replied:
Done.

bclary commented on taskcluster/script.py:
```diff
@@ -203,22 +261,32 @@ def main():
+                temp_bytes_written = 0
+                while temp_bytes_written != line_len:
+                    temp_bytes_written += sys.stdout.write(decoded_line)
+                    if temp_bytes_written != line_len:
```
ditto <=

aerickson replied:
I think that this could hide the first underwrite if it the write sends it out before the print.

Done.

bclary reviewed (changes requested):
Let's make the <= changes. I definitely don't want this merged to master though. Let's just deploy this branch to a testing image to get the debugging information and we can move on from there. I'd like to schedule a meeting this coming week to talk about this script.py. I think we should consider making a number of changes but I'd like to work more closely with you on them.

aerickson committed:
remove vim, fix trailing newline

aerickson committed:
change != to <=

aerickson commented:
> Let's make the <= changes. I definitely don't want this merged to master though. Let's just deploy this branch to a testing image to get the debugging information and we can move on from there. I'd like to schedule a meeting this coming week to talk about this script.py. I think we should consider making a number of changes but I'd like to work more closely with you on them.

Yeah, the previous change wasn't landed to master either.

Meeting sounds good.

bclary reviewed (approved):
r+. Let's try it out in testing.

aerickson commented:
https://github.com/bclary/mozilla-bitbar-docker/pull/41/commits/d0bd089514e32aedec2cf76401b35e8585470450 broke the logic from looking at the try output.

aerickson committed:
<= to <

aerickson committed:
increase interval for logcat printing, also show pstree

aerickson commented:
Tested locally and doesn't do the double write or false underwrite notification like in d0bd089 (https://treeherder.mozilla.org/#/jobs?repo=try&revision=eab0ca30d81b23a22eff475737735971dbc58039&selectedJob=288254169).

Hopefully the added pstree calls in the interval logcat logging will help identify what's going on.

changes since last review:
  - https://github.com/bclary/mozilla-bitbar-docker/pull/41/files/d0bd089514e32aedec2cf76401b35e8585470450..87f28ae93802ce62ca2c4140bba17ab669df5410

bclary reviewed (approved):
r+

aerickson committed:
stop printing to logcat, name refactoring

causing ADBTimeout exceptions when logging every 20 seconds

aerickson commented:
The most recent try job is failing when logging to logcat. pstree to logcat every 20 seconds seems to be too much. 
- https://treeherder.mozilla.org/#/jobs?repo=try&revision=0614122c79c5c2f545bc21aad06e53f800e8bb32&selectedJob=288275640

It's disturbing that the devices are so fragile. I wouldn't think a few log calls would cause the device to crash.

aerickson commented:
Building docker image 20200211T124654 based off of 0f3a6f8.

https://mozilla.testdroid.com/#testing/device-session/1613876/1701908/45573288
- `02-11 12:51:56.955 Successfully tagged mozilla-image:20200211T124654`

test results
- https://treeherder.mozilla.org/#/jobs?repo=try&revision=cb93186bd82c0e32531e7be7e0dc7f38988381ad&group_state=expanded&selectedJob=288418626
  - seems like script.py is hanging
    - next idea: reduce things that gate exiting the loop

aerickson committed:
if timed out, exit 1

aerickson committed:
close_fds, simplify print logic

aerickson committed:
fix extra newline

aerickson committed:
set shorter timeout (save testing time)

aerickson committed:
fetch rc

aerickson commented:
Building docker image 20200211T182130 from f18815b.

`02-11 18:26:08.829 Successfully tagged mozilla-image:20200211T182130`

tests:
- https://treeherder.mozilla.org/#/jobs?repo=try&revision=d6c967df0b19bf76e84ddfe307c1f40195b8226a
  - still hanging. not waiting on proc.poll()/rc.

aerickson committed:
add a non-blocking attempt

aerickson commented:
Building 20200212T113800 based off of 1644920.
- `02-12 11:42:25.232 Successfully tagged mozilla-image:20200212T113800`

Tests:
- https://treeherder.mozilla.org/#/jobs?repo=try&revision=b322f716e3c855cce85df1197ef35189876d88a3
  - works!!!
    - why? i think it was a race between proc.poll and readline. the process returned, but readline was already waiting for output.

Next ideas if this doesn't work:
- get rid of readline, still not sure we're not hanging on waiting for a newline

aerickson committed:
add note

aerickson commented:
Fix found. Creating a PR that will land on master in #42.
